### ISSUE TITLE
Many params of interface methods incorrectly marked as "optional"

### ISSUE THREAD

**Issue by UselessPickles** [ag-grid/ag-grid#1979]
**I'm submitting a ...**  (check one with "x")
```
[X] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/ceolter/ag-grid/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
When using the "strict" compiler option for the TypeScript compiler, I have to mark params as "optional" in many interface method implementations where it makes no sense at all that the param be optional. For example, the "event" param in ALL event handler methods is marked as "optional".

Excerpt from gridOptions.ts:
{code}
    // events
    onColumnEverythingChanged?(event?: ColumnEverythingChangedEvent): void;
    onNewColumnsLoaded?(event?: NewColumnsLoadedEvent): void;
    onColumnPivotModeChanged?(event?: ColumnPivotModeChangedEvent): void;
    onColumnRowGroupChanged?(event?: ColumnRowGroupChangedEvent): void;
    onColumnPivotChanged?(event?: ColumnPivotChangedEvent): void;
    onGridColumnsChanged?(event?: GridColumnsChangedEvent): void;
    onColumnValueChanged?(event?: ColumnValueChangedEvent): void;
    onColumnMoved?(event?: ColumnMovedEvent): void;
    onColumnVisible?(event?: ColumnVisibleEvent): void;
    onColumnPinned?(event?: ColumnPinnedEvent): void;
    onColumnGroupOpened?(event?: ColumnGroupOpenedEvent): void;
    onColumnResized?(event?: ColumnResizedEvent): void;
{code}

Because the "event" is optional in the interface signature, the TypeScript compiler assumes that its value may be undefined. With "strict" compiler options, I am then forced to perform a null-check in my implementations of all these events before accessing any properties of the event. The code then appears awkward because it looks like I'm not fully implementing the event handler - I'm not doing anything to handle the case where "event" is undefined/null. What am I supposed to do in that case?

This appears to be a widespread issue in the code base in nearly all interface methods. Why are params unnecessarily marked as "optional"?

NOTE: this is also somewhat related to this other issue caused by an interface method signature with a param unnecessarily marked as "optional".

**Expected behavior**
Interface methods should not have params marked as optional when the intent is for a non-null/non-undefined value to always be passed to the implementation. I should not have to perform an awkward null check on a param that is never actually expected to be null/undefined.

**Minimal reproduction of the problem with instructions**
Simply use the TypeScript compiler's "strict" option when compiling any project that makes use of any of the grid's event handlers. The compiler will throw an error if the "event" param is not marked as optional in your implementation, then after marking it optional, all attempts to access properties of "event" will cause an error because "event" may be undefined. You'll then need to surround the code with a null check.

Example implementation of an event handler (broken, but this is what SHOULD WORK):
{code}
    public render(): JSXElement {
        // Compiler error at this.myColumnResizedHandler(event) because event may be undefined,
        // but it is not marked as optional or potentially undefined in the signature
        // of myColumnResizedHandler
        return (
            <AgGridReact onColumnResized={(event) => this.myColumnResizedHandler(event)} />
        );
    }

    private myColumnResizedHandler(event: ColumnResizedEvent): void {
        console.log(event.finished);
    }
{code}

Workaround that is currently necessary to pass compilation:
{code}
    public render(): JSXElement {
        return (
            <AgGridReact onColumnResized={(event) => this.myColumnResizedHandler(event)} />
        );
    }

    // "event" marked as optional here
    private myColumnResizedHandler(event?: ColumnResizedEvent): void {
        // Null check required to prevent compiler error on next line
        if (event) {
            console.log(event.finished);
        }
    }
{code}

**What is the motivation / use case for changing the behavior?**
I want to compile my TypeScript project with "strict" compiler option and not be forced to work around incorrect method signatures in ag-grid's code.

**Please tell us about your environment:**
Mostly irrelevant, except for using TypeScript with the "strict" compiler option enabled. Most specifically, the "strict-null-checks" option reveals this issue.

* **ag-Grid version:** 13.3.1

* **Browser:** Irrelevant

* **Language:** TypeScript 2.5.3

**Comment by amirburbea** [comment]
Agreed - this manifests in a couple other places as well (now even in trying to just build apps using newer builds of ag-grid when strictNullChecks is on).

As a workaround, if you're working with an older build where you can at least compile with ag-grid, you can use an exclamation mark to override the null check as the following

```ts
public render(): JSX.Element {
        return (
            <AgGridReact onColumnResized={(event) => this.myColumnResizedHandler(event!)} />
        );
    }

    // "event" marked as optional here
    private myColumnResizedHandler(event: ColumnResizedEvent): void {
       console.log(event.finished);
    }
```

**Comment by UselessPickles** [comment]
I forked this repo, enabled the "strict" compiler option, and started working through all of the null/undefined issues in the ag-grid source code. 

One major issue that I can't solve without assistance is the Component.getHtmlElement() method in src/ts/widgets/component.ts. The method returns null if called " before gui was attached". Changing the signature to support returning null quickly reveals that many usages of the method simply assume that the result is non-null. This seems like a situation where getHtmlElement() should throw an error instead of returning null. Otherwise, several methods will need too perform null checks, and in turn return null if the getHtmlElement() is null. Then all callers of all those methods will have to be updated to handle a null result. This is a massive ripple effect for handling a critical error condition.

Can I get a recommendation from someone familiar with the project on whether it would be appropriate for getHtmlElement() to throw an error instead of returning null? Is that really an error condition that is intended to be "recoverable" by returning null? If not, then throwing an error would greatly simplify the code by guaranteeing the method never returns null.

Or is it more appropriate to update some of the usages of getHtmlElement() that currently ASSUME a non-null result to perform a null check and throw an error?

It's hard for me to know what's right without being intimately familiar with the design of this code.

**Comment by UselessPickles** [comment]
This is getting pretty hopeless for me to fix myself. For example, RowNode.id, RowNode.rowHeight, etc, are set to null in some situations. There is plenty of code that assumes those properties are not null. how to fix the code to deal with those properties being potentially null is unclear, especially when there are mathematical calculations/comparisons involving RowNode.rowHeight, which may be null, but the existing code ignores that possibility.

I think someone that understands this codebase well needs to simply enable the "strict" compiler option, then work through all the compiler errors a few at a time. You will discover some bad assumptions in the code, and will need to make non-trivial decisions about how to handle null/undefined values. There's probably a few nasty bugs hiding in there that will be fixed by updating the code to pass a "strict" compile.

Also, FYI, the _.exists() method is quite useless, because the compiler does not recognize it as guaranteeing that the param is not null. In most cases, at least for non-string types, it's much better to simply test the variable directly (test if == or != null, with double instead of triple equal to handle both null and undefined) without using _.exists() so that the compiler is aware that you performed a null check.

**Comment by amirburbea** [comment]
If you know a property is actually not null/undefined in some context, you
can simply use ! (As in instance!.property).

I'll fork later and give it a shot.

**Comment by UselessPickles** [comment]
> If you know a property is actually not null/undefined in some context, you
can simply use ! (As in instance!.property).

That should really only be used sparingly in situations where you are 100% sure that it is (and always should be after future code refactors/changes) impossible to reach that code in a way that the property/param/var is null. It's not a general solution to this problem.

The general problem is that this code base has completely lost control/knowledge of when things are allowed to be null and what it means, how to behave when things are null. 

If the goal was to simply make the code pass "strict" compilation, then we could scatter usages of "!" everywhere to just make the compiler stop complaining. But the whole point of strict compilation is that it forces you to be aware of when you are dealing with possibly null things, and at every single place where a value may be null, a proper decision needs to be made as to how to handle  the possibly null value. The compiler "strict" compiler errors often reveal potential bugs.
